File: joka/infra/fs.py
import re

from pathlib import Path
from typing import List
from datetime import datetime

from joka.files.models import MigrationFile

MIGRATION_PATTERN = re.compile(r"^(\d{12})_.*\.sql$")  # yymmddhhmmss_*.sql

# exception: migration directory was missing
class MigrationDirectoryNotFoundError(Exception):
    """Raised when the specified migrations directory does not exist."""
    pass

# exception: migration dir not found
class MigrationDirNotFoundError(Exception):
    """Raised when the specified migrations directory does not exist."""
    pass


def create_migration_file(migration_dir: str, migration_name: str) -> None:
    """
    Create a new migration file in the specified directory with the given name.
    """

    dir_path = Path(migration_dir)

    # check if path exists or throw exception
    if not dir_path.exists():
        raise MigrationDirNotFoundError(f"Migrations directory not found: {migration_dir}")

    # create a new migration file with a timestamp
    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
    migration_filename = f"{timestamp}_{migration_name}.sql"
    migration_filepath = dir_path / migration_filename

    with migration_filepath.open("w") as f:
        f.write("-- Write your migration SQL here\n")


def list_migration_files(dir: str) -> List[MigrationFile]:
    """
    List all migration files in the given directory, sorted by name.
    """

    dir_path = Path(dir)

    # ensure the path exists
    if not dir_path.exists() or not dir_path.is_dir():
        raise MigrationDirNotFoundError(f"Migrations directory not found: {dir}")
    
    # make a list of MigrationFile
    migration_files: List[MigrationFile] = []

    for f in dir_path.iterdir():
        if f.is_file() and f.suffix == ".sql":
            match = MIGRATION_PATTERN.match(f.name)
            if match:
                index = match.group(1)
                name = f.name[len(index) + 1:-4]  # remove index and .sql
                migration_files.append(
                    MigrationFile(
                        index=index,
                        name=name,
                        full_path=str(f.resolve())
                    )
                )

    # sort by index
    migration_files.sort(key=lambda mf: mf.index)

    return migration_files

File: internal/domains/migration/infra/fs.go
package infra

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"time"

	"github.com/nickfiggins/joka/internal/domains/migration/infra/models"
)

// migrationPattern matches filenames of the form YYMMDDHHMMSS_name.sql.
var migrationPattern = regexp.MustCompile(`^(\d{12})_.*\.sql$`)

// ListMigrationFiles scans dir for SQL files matching the migration naming
// convention and returns them sorted by their timestamp index. Non-matching
// files and subdirectories are silently ignored.
func ListMigrationFiles(dir string) ([]models.MigrationFile, error) {
	info, err := os.Stat(dir)
	if err != nil || !info.IsDir() {
		return nil, fmt.Errorf("migrations directory not found: %s", dir)
	}

	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, fmt.Errorf("reading migrations directory: %w", err)
	}

	var files []models.MigrationFile
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		matches := migrationPattern.FindStringSubmatch(name)
		if matches == nil {
			continue
		}
		index := matches[1]
		// name is everything after the index + underscore, minus .sql
		migName := name[len(index)+1 : len(name)-4]
		fullPath, _ := filepath.Abs(filepath.Join(dir, name))

		files = append(files, models.MigrationFile{
			Index:    index,
			Name:     migName,
			FullPath: fullPath,
		})
	}

	sort.Slice(files, func(i, j int) bool {
		return files[i].Index < files[j].Index
	})

	return files, nil
}

// CreateMigrationFile creates a new empty SQL migration file in dir using the
// current timestamp as a prefix. It returns the generated filename (not the
// full path). The migrations directory must already exist.
func CreateMigrationFile(dir string, name string) (string, error) {
	info, err := os.Stat(dir)
	if err != nil || !info.IsDir() {
		return "", fmt.Errorf("migrations directory not found: %s", dir)
	}

	timestamp := time.Now().Format("20060102150405")
	filename := fmt.Sprintf("%s_%s.sql", timestamp, name)
	filePath := filepath.Join(dir, filename)

	if err := os.WriteFile(filePath, []byte("-- Write your migration SQL here\n"), 0644); err != nil {
		return "", fmt.Errorf("creating migration file: %w", err)
	}

	return filename, nil
}
